File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiHeatmapLayer.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Shapes;
using Windows.UI;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Heatmap Layer data plot (unified-component-catalog.md
    /// §2.6, "Heatmap Layer" row: "Density overlay synced to a camera",
    /// built from _(plot primitive)_) — a density grid rendered as an
    /// alpha-blended overlay (e.g. a face-detection density map synced to a
    /// map/photo viewport camera elsewhere in the app). The camera-sync
    /// itself is the host's concern — matching `mui-heatmap-layer.component.ts`'s
    /// own scope, this control only rasterizes the grid it's given. Cell
    /// color reads the accent token, alpha-blended per cell by that cell's
    /// normalized density.
    /// </summary>
    public sealed class MuiHeatmapLayer : ContentControl
    {
        public static readonly DependencyProperty DensityProperty =
            DependencyProperty.Register(nameof(Density), typeof(IReadOnlyList<IReadOnlyList<double>>), typeof(MuiHeatmapLayer),
                new PropertyMetadata(null, (d, _) => ((MuiHeatmapLayer)d).Render()));

        public static readonly DependencyProperty PlotWidthProperty =
            DependencyProperty.Register(nameof(PlotWidth), typeof(double), typeof(MuiHeatmapLayer),
                new PropertyMetadata(160.0, (d, _) => ((MuiHeatmapLayer)d).Rebuild()));

        public static readonly DependencyProperty PlotHeightProperty =
            DependencyProperty.Register(nameof(PlotHeight), typeof(double), typeof(MuiHeatmapLayer),
                new PropertyMetadata(96.0, (d, _) => ((MuiHeatmapLayer)d).Rebuild()));

        /// <summary>Rows of per-cell density, each 0..1. Every row must be
        /// the same length; an empty grid draws nothing.</summary>
        public IReadOnlyList<IReadOnlyList<double>>? Density
        {
            get => (IReadOnlyList<IReadOnlyList<double>>?)GetValue(DensityProperty);
            set => SetValue(DensityProperty, value);
        }

        public double PlotWidth
        {
            get => (double)GetValue(PlotWidthProperty);
            set => SetValue(PlotWidthProperty, value);
        }

        public double PlotHeight
        {
            get => (double)GetValue(PlotHeightProperty);
            set => SetValue(PlotHeightProperty, value);
        }

        private readonly Grid _frame = new() { CornerRadius = new CornerRadius(6), BorderThickness = new Thickness(0.5) };
        private readonly Canvas _canvas = new();

        public MuiHeatmapLayer()
        {
            _frame.Children.Add(_canvas);
            Content = _frame;
            IsTabStop = false;

            Rebuild();
        }

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void Rebuild()
        {
            _frame.Width = PlotWidth;
            _frame.Height = PlotHeight;
            _canvas.Width = PlotWidth;
            _canvas.Height = PlotHeight;
            _frame.Background = R("MapleImageCanvas");
            _frame.BorderBrush = R("MapleBorder");
            Render();
        }

        private void Render()
        {
            _canvas.Children.Clear();
            var w = PlotWidth;
            var h = PlotHeight;
            var rows = Density ?? Array.Empty<IReadOnlyList<double>>();
            if (w <= 0 || h <= 0 || rows.Count == 0 || rows[0].Count == 0) return;

            var cellWidth = MuiHeatmapMath.CellWidth(rows, w);
            var cellHeight = MuiHeatmapMath.CellHeight(rows, h);
            var baseColor = ((SolidColorBrush)R("MaplePrimary")).Color;

            for (var rowIndex = 0; rowIndex < rows.Count; rowIndex++)
            {
                var row = rows[rowIndex];
                for (var colIndex = 0; colIndex < row.Count; colIndex++)
                {
                    var density = MuiHeatmapMath.ClampDensity(row[colIndex]);
                    if (density <= 0) continue;

                    var alpha = (byte)Math.Round(density * 255);
                    var rect = new Rectangle
                    {
                        Width = cellWidth,
                        Height = cellHeight,
                        Fill = new SolidColorBrush(Color.FromArgb(alpha, baseColor.R, baseColor.G, baseColor.B)),
                    };
                    Canvas.SetLeft(rect, colIndex * cellWidth);
                    Canvas.SetTop(rect, rowIndex * cellHeight);
                    _canvas.Children.Add(rect);
                }
            }
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiHeatmapLayer.swift
// MuiHeatmapLayer.swift — Maple UI Molecules-L1 (unified-component-
// catalog.md §2.6; a plot primitive). A density grid rendered as an
// alpha-blended overlay via SwiftUI `Canvas` (e.g. a face-detection
// density map synced to a map/photo viewport elsewhere in the app — the
// camera-sync itself is the host's concern; this component only rasterizes
// the grid it's given). Cell color reads the accent token, alpha-blended
// per cell by that cell's normalized density.

import SwiftUI

public struct MuiHeatmapLayer: View {
    /// Rows of per-cell density, each `0...1`. Every row must be the same
    /// length; an empty grid draws nothing.
    public let grid: [[Double]]
    public let width: CGFloat
    public let height: CGFloat
    public let color: Color

    public init(grid: [[Double]], width: CGFloat = 160, height: CGFloat = 96, color: Color = MuiTokens.primary) {
        self.grid = grid
        self.width = width
        self.height = height
        self.color = color
    }

    public var body: some View {
        Canvas { context, size in
            guard let firstRow = grid.first, !firstRow.isEmpty else { return }
            let cols = firstRow.count
            let cellW = size.width / CGFloat(cols)
            let cellH = size.height / CGFloat(grid.count)

            for (rowIndex, row) in grid.enumerated() {
                for (colIndex, value) in row.enumerated() {
                    let density = MuiCurvePlotMath.clampUnit(value)
                    guard density > 0 else { continue }
                    let rect = CGRect(x: CGFloat(colIndex) * cellW, y: CGFloat(rowIndex) * cellH, width: cellW, height: cellH)
                    context.fill(Path(rect), with: .color(color.opacity(density)))
                }
            }
        }
        .frame(width: width, height: height)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Heatmap layer")
    }
}

#Preview("MuiHeatmapLayer") {
    let grid: [[Double]] = (0..<8).map { row -> [Double] in
        (0..<8).map { col -> Double in
            let distance: Double = Double(abs(row - 4) + abs(col - 4))
            return Swift.max(0, 1 - distance / 6.0)
        }
    }
    return MuiHeatmapLayer(grid: grid)
        .padding()
        .background(MuiTokens.imageCanvas)
}
